File: src/agents/master-decision-engine.ts
// ============================================================================
// MASTER DECISION ENGINE
// No single indicator may generate signals alone.
// Every signal requires confirmation from multiple modules.
//
// Required confluence for a valid signal:
//   Liquidity Sweep + BOS + FVG + Order Block + SMT + Session + HTF Bias +
//   Trend + Volume => Generate Buy/Sell Signal with Confidence Score
// ============================================================================

import { Direction, Bias } from '../core/types';
import { AgentOutput, AgentName, OrchestratorResult } from './types';

export type SignalGrade =
  | 'NO_SIGNAL'
  | 'WEAK'
  | 'MODERATE'
  | 'STRONG'
  | 'VERY_STRONG'
  | 'INSTITUTIONAL';

/** Required confluences that MUST be present for signal approval */
export type RequiredConfluence =
  | 'LIQUIDITY_SWEEP'
  | 'BOS_OR_CHOCH'
  | 'FVG'
  | 'ORDER_BLOCK'
  | 'SMT'
  | 'SESSION'
  | 'HTF_BIAS'
  | 'TREND'
  | 'VOLUME';


export interface DecisionConfig {
  /** Minimum required confluences to approve a signal (from the 9 above) */
  minRequiredConfluences: number;
  /** Minimum confidence score from the Strategy Agent */
  minStrategyConfidence: number;
  /** If Risk or Psychology agent blocks, override all */
  respectRiskBlock: boolean;
  /** Minimum total weighted score from all agents */
  minWeightedScore: number;
}

const DEFAULT_CONFIG: DecisionConfig = {
  minRequiredConfluences: 5, // At least 5 of 9 required confluences
  minStrategyConfidence: 55,
  respectRiskBlock: true,
  minWeightedScore: 50,
};

export interface DecisionResult {
  approved: boolean;
  direction: Direction | null;
  confidence: number;
  grade: SignalGrade;
  /** Which required confluences are present */
  confluencePresent: RequiredConfluence[];
  /** Which are missing */
  confluenceMissing: RequiredConfluence[];
  /** Blocking reasons (if not approved) */
  blockReasons: string[];
  /** Narrative explanation */
  explanation: string;
  timestamp: number;
}

export class MasterDecisionEngine {
  private config: DecisionConfig;

  constructor(config: Partial<DecisionConfig> = {}) {
    this.config = { ...DEFAULT_CONFIG, ...config };
  }

  /**
   * Evaluate an orchestrator result and produce a final go/no-go decision.
   * This is the ULTIMATE gatekeeper — no trade without its approval.
   */
  evaluate(result: OrchestratorResult): DecisionResult {
    const { agentOutputs, aggregateConfidence, signalDirection } = result;

    // 1. Check Risk/Psychology blocks
    const blockReasons: string[] = [];
    if (this.config.respectRiskBlock) {
      const riskOut = agentOutputs.get('RISK');
      const psychOut = agentOutputs.get('PSYCHOLOGY');

      if (riskOut?.insights.some(i => i.tags.includes('BLOCK'))) {
        blockReasons.push(`Risk Agent BLOCKED: ${riskOut.insights.filter(i => i.tags.includes('BLOCK')).map(i => i.type).join(', ')}`);
      }
      if (psychOut?.insights.some(i => i.tags.includes('BLOCK'))) {
        blockReasons.push(`Psychology Agent BLOCKED: ${psychOut.insights.filter(i => i.tags.includes('BLOCK')).map(i => i.type).join(', ')}`);
      }
    }

    if (blockReasons.length > 0) {
      return {
        approved: false, direction: null, confidence: 0, grade: 'NO_SIGNAL',
        confluencePresent: [], confluenceMissing: this.allConfluences(),
        blockReasons,
        explanation: `DECISION: NO TRADE. ${blockReasons.join(' | ')}`,
        timestamp: Date.now(),
      };
    }

    // 2. If no direction from orchestrator, no signal
    if (!signalDirection) {
      return {
        approved: false, direction: null, confidence: aggregateConfidence, grade: 'NO_SIGNAL',
        confluencePresent: [], confluenceMissing: this.allConfluences(),
        blockReasons: ['No directional consensus from agents'],
        explanation: 'DECISION: No consensus — stand aside.',
        timestamp: Date.now(),
      };
    }

    // 3. Check required confluences
    const present: RequiredConfluence[] = [];
    const missing: RequiredConfluence[] = [];
    const allInsights = Array.from(agentOutputs.values()).flatMap(o => o.insights);
    const dirInsights = allInsights.filter(i => i.direction === signalDirection);

    // Check each required confluence
    this.checkConfluence('LIQUIDITY_SWEEP', dirInsights, present, missing, ['LIQUIDITY_SWEEP', 'SWEEP']);
    this.checkConfluence('BOS_OR_CHOCH', dirInsights, present, missing, ['BOS', 'CHOCH', 'MSS']);
    this.checkConfluence('FVG', dirInsights, present, missing, ['FVG', 'IFVG', 'BPR']);
    this.checkConfluence('ORDER_BLOCK', dirInsights, present, missing, ['BULLISH_OB', 'BEARISH_OB', 'BREAKER', 'MITIGATION', 'ORDER_BLOCK']);
    this.checkConfluence('SMT', dirInsights, present, missing, ['SMT']);

    // Session from ICT agent
    const ictOut = agentOutputs.get('ICT');
    if (ictOut?.insights.some(i => i.type === 'KILL_ZONE')) {
      present.push('SESSION');
    } else { missing.push('SESSION'); }

    // HTF Bias — check if Market Structure agent agrees with direction
    const structOut = agentOutputs.get('MARKET_STRUCTURE');
    if (structOut?.bias === signalDirection) { present.push('HTF_BIAS'); } else { missing.push('HTF_BIAS'); }

    // Trend — check trend agent alignment
    const trendOut = agentOutputs.get('TREND');
    if (trendOut?.bias === signalDirection) { present.push('TREND'); } else { missing.push('TREND'); }

    // Volume — check volume agent delta alignment
    const volOut = agentOutputs.get('VOLUME');
    const deltaInsight = volOut?.insights.find(i => i.type === 'DELTA');
    if (deltaInsight?.direction === signalDirection) { present.push('VOLUME'); } else { missing.push('VOLUME'); }

    // 4. Decision
    const meetsConfluence = present.length >= this.config.minRequiredConfluences;
    const meetsConfidence = aggregateConfidence >= this.config.minStrategyConfidence;
    const approved = meetsConfluence && meetsConfidence;

    if (!meetsConfluence) {
      blockReasons.push(`Only ${present.length}/${this.config.minRequiredConfluences} required confluences present. Missing: ${missing.join(', ')}`);
    }
    if (!meetsConfidence) {
      blockReasons.push(`Confidence ${aggregateConfidence}% below minimum ${this.config.minStrategyConfidence}%`);
    }

    const grade = this.gradeSignal(present.length, aggregateConfidence);

    return {
      approved, direction: approved ? signalDirection : null,
      confidence: aggregateConfidence, grade,
      confluencePresent: present, confluenceMissing: missing,
      blockReasons,
      explanation: approved
        ? `DECISION: ${signalDirection} APPROVED (${grade}) — ${present.length}/9 confluences, ${aggregateConfidence}% confidence.`
        : `DECISION: REJECTED — ${blockReasons.join('; ')}`,
      timestamp: Date.now(),
    };
  }

  private checkConfluence(
    name: RequiredConfluence,
    insights: { type: string; tags: string[] }[],
    present: RequiredConfluence[],
    missing: RequiredConfluence[],
    matchTypes: string[]
  ): void {
    const found = insights.some(i =>
      matchTypes.includes(i.type) || i.tags.some(t => matchTypes.includes(t))
    );
    if (found) present.push(name); else missing.push(name);
  }

  private gradeSignal(confluenceCount: number, confidence: number): SignalGrade {
    if (confluenceCount >= 8 && confidence >= 85) return 'INSTITUTIONAL';
    if (confluenceCount >= 7 && confidence >= 75) return 'VERY_STRONG';
    if (confluenceCount >= 6 && confidence >= 65) return 'STRONG';
    if (confluenceCount >= 5 && confidence >= 55) return 'MODERATE';
    if (confluenceCount >= 4) return 'WEAK';
    return 'NO_SIGNAL';
  }

  private allConfluences(): RequiredConfluence[] {
    return ['LIQUIDITY_SWEEP', 'BOS_OR_CHOCH', 'FVG', 'ORDER_BLOCK', 'SMT', 'SESSION', 'HTF_BIAS', 'TREND', 'VOLUME'];
  }

  updateConfig(config: Partial<DecisionConfig>): void {
    this.config = { ...this.config, ...config };
  }
}

File: app/src/main/java/com/foxtrader/app/domain/usecase/ai/MasterDecisionEngine.kt
package com.foxtrader.app.domain.usecase.ai

import com.foxtrader.app.domain.model.AgentInsight
import com.foxtrader.app.domain.model.AgentName
import com.foxtrader.app.domain.model.Bias
import com.foxtrader.app.domain.model.DecisionConfig
import com.foxtrader.app.domain.model.DecisionResult
import com.foxtrader.app.domain.model.Direction
import com.foxtrader.app.domain.model.OrchestratorResult
import com.foxtrader.app.domain.model.RequiredConfluence
import com.foxtrader.app.domain.model.SignalGrade
import javax.inject.Inject

/**
 * MASTER DECISION ENGINE — the ultimate gatekeeper.
 *
 * No single agent may authorize a trade. A signal is approved only when:
 *  1. Neither the RISK nor PSYCHOLOGY agent raised a BLOCK (absolute veto), AND
 *  2. The orchestrator produced a directional signal, AND
 *  3. At least [DecisionConfig.minRequiredConfluences] of the 9 institutional
 *     confluences are present in the signal's direction, AND
 *  4. Aggregate confidence >= [DecisionConfig.minConfidence].
 *
 * Kotlin port of the reference `master-decision-engine.ts`. Pure and
 * deterministic given an [OrchestratorResult].
 */
class MasterDecisionEngine @Inject constructor() {

    private var config: DecisionConfig = DecisionConfig()

    fun updateConfig(newConfig: DecisionConfig) { config = newConfig }
    fun getConfig(): DecisionConfig = config

    fun evaluate(result: OrchestratorResult): DecisionResult {
        val now = System.currentTimeMillis()
        val outputs = result.agentOutputs

        // 1. Risk / Psychology veto (absolute).
        if (config.respectRiskBlock) {
            val riskBlock = outputs[AgentName.RISK]?.insights?.filter { it.isBlock() }.orEmpty()
            val psychBlock = outputs[AgentName.PSYCHOLOGY]?.insights?.filter { it.isBlock() }.orEmpty()
            val veto = when {
                riskBlock.isNotEmpty() -> AgentName.RISK to riskBlock
                psychBlock.isNotEmpty() -> AgentName.PSYCHOLOGY to psychBlock
                else -> null
            }
            if (veto != null) {
                val (who, blocks) = veto
                val reasons = blocks.map { it.detail }
                return DecisionResult(
                    approved = false,
                    direction = null,
                    confidence = 0.0,
                    grade = SignalGrade.NO_SIGNAL,
                    confluencePresent = emptyList(),
                    confluenceMissing = RequiredConfluence.all(),
                    blockReasons = listOf("$who BLOCKED: ${reasons.joinToString(", ")}"),
                    vetoedBy = who,
                    explanation = "DECISION: NO TRADE. $who veto — ${reasons.joinToString("; ")}",
                    timestamp = now,
                )
            }
        }

        // 2. No directional consensus -> stand aside.
        val direction = result.signalDirection
        if (direction == null) {
            return DecisionResult(
                approved = false,
                direction = null,
                confidence = result.aggregateConfidence,
                grade = SignalGrade.NO_SIGNAL,
                confluencePresent = emptyList(),
                confluenceMissing = RequiredConfluence.all(),
                blockReasons = listOf("No directional consensus from agents"),
                vetoedBy = null,
                explanation = "DECISION: No consensus — stand aside.",
                timestamp = now,
            )
        }

        // 3. Required-confluence check.
        val allInsights = outputs.values.flatMap { it.insights }
        val dirInsights = allInsights.filter { it.direction == direction }

        val present = mutableListOf<RequiredConfluence>()
        val missing = mutableListOf<RequiredConfluence>()

        checkConfluence(RequiredConfluence.LIQUIDITY_SWEEP, dirInsights, present, missing,
            setOf("LIQUIDITY_SWEEP", "SWEEP"))
        checkConfluence(RequiredConfluence.BOS_OR_CHOCH, dirInsights, present, missing,
            setOf("BOS", "CHOCH", "MSS"))
        checkConfluence(RequiredConfluence.FVG, dirInsights, present, missing,
            setOf("FVG", "IFVG", "BPR"))
        checkConfluence(RequiredConfluence.ORDER_BLOCK, dirInsights, present, missing,
            setOf("BULLISH_OB", "BEARISH_OB", "BREAKER", "MITIGATION", "ORDER_BLOCK"))
        checkConfluence(RequiredConfluence.SMT, dirInsights, present, missing, setOf("SMT"))

        // SESSION — kill zone present (direction-agnostic).
        if (allInsights.any { it.type == "KILL_ZONE" || it.tags.contains("SESSION") }) {
            present += RequiredConfluence.SESSION
        } else {
            missing += RequiredConfluence.SESSION
        }

        // HTF_BIAS — Market Structure agent agrees with the signal direction.
        if (outputs[AgentName.MARKET_STRUCTURE]?.bias.agreesWith(direction)) {
            present += RequiredConfluence.HTF_BIAS
        } else {
            missing += RequiredConfluence.HTF_BIAS
        }

        // TREND — Trend agent agrees with the signal direction.
        if (outputs[AgentName.TREND]?.bias.agreesWith(direction)) {
            present += RequiredConfluence.TREND
        } else {
            missing += RequiredConfluence.TREND
        }

        // VOLUME — a DELTA insight aligned with the signal direction.
        val deltaAligned = outputs[AgentName.VOLUME]?.insights
            ?.any { it.type == "DELTA" && it.direction == direction } == true
        if (deltaAligned) present += RequiredConfluence.VOLUME else missing += RequiredConfluence.VOLUME

        // 4. Final decision.
        val meetsConfluence = present.size >= config.minRequiredConfluences
        val meetsConfidence = result.aggregateConfidence >= config.minConfidence
        val approved = meetsConfluence && meetsConfidence

        val blockReasons = mutableListOf<String>()
        if (!meetsConfluence) {
            blockReasons += "Only ${present.size}/${config.minRequiredConfluences} required " +
                "confluences present. Missing: ${missing.joinToString(", ")}"
        }
        if (!meetsConfidence) {
            blockReasons += "Confidence ${result.aggregateConfidence.toInt()}% below minimum " +
                "${config.minConfidence.toInt()}%"
        }

        val grade = gradeSignal(present.size, result.aggregateConfidence)

        return DecisionResult(
            approved = approved,
            direction = if (approved) direction else null,
            confidence = result.aggregateConfidence,
            grade = if (approved) grade else SignalGrade.NO_SIGNAL,
            confluencePresent = present,
            confluenceMissing = missing,
            blockReasons = blockReasons,
            vetoedBy = null,
            explanation = if (approved) {
                "DECISION: $direction APPROVED ($grade) — ${present.size}/9 confluences, " +
                    "${result.aggregateConfidence.toInt()}% confidence."
            } else {
                "DECISION: REJECTED — ${blockReasons.joinToString("; ")}"
            },
            timestamp = now,
        )
    }

    private fun checkConfluence(
        confluence: RequiredConfluence,
        insights: List<AgentInsight>,
        present: MutableList<RequiredConfluence>,
        missing: MutableList<RequiredConfluence>,
        matchTypes: Set<String>,
    ) {
        val found = insights.any { it.type in matchTypes || it.tags.any { tag -> tag in matchTypes } }
        if (found) present += confluence else missing += confluence
    }

    private fun gradeSignal(confluenceCount: Int, confidence: Double): SignalGrade = when {
        confluenceCount >= 8 && confidence >= 85 -> SignalGrade.INSTITUTIONAL
        confluenceCount >= 7 && confidence >= 75 -> SignalGrade.VERY_STRONG
        confluenceCount >= 6 && confidence >= 65 -> SignalGrade.STRONG
        confluenceCount >= 5 && confidence >= 55 -> SignalGrade.MODERATE
        confluenceCount >= 4 -> SignalGrade.WEAK
        else -> SignalGrade.NO_SIGNAL
    }

    private fun AgentInsight.isBlock(): Boolean = tags.contains("BLOCK")

    /** Whether a (nullable) [Bias] agrees with a trade [Direction]. */
    private fun Bias?.agreesWith(direction: Direction): Boolean = when (this) {
        Bias.BULLISH -> direction == Direction.BULLISH
        Bias.BEARISH -> direction == Direction.BEARISH
        else -> false
    }
}
